webextension/background.js

```diff
@@ -1,3 +1,206 @@
+    area: browser.storage.local,
+
+    getOriginStoreKey(pageUrl) {
+      const url = new window.URL(pageUrl);
+      const storagePrefix = "originContainerMap@@_";
+      return `${storagePrefix}${url.origin}`;
+    },
+
+    get(pageUrl) {
+      const originStoreKey = this.getOriginStoreKey(pageUrl);
+      return new Promise((resolve, reject) => {
+        this.area.get([originStoreKey]).then((storageResponse) => {
+          if (storageResponse && originStoreKey in storageResponse) {
+            resolve(storageResponse[originStoreKey]);
+          }
+          resolve(null);
+        }).catch(() => {
+          reject();
+        });
+      });
+    },
+
+    set(pageUrl, data) {
+      const originStoreKey = this.getOriginStoreKey(pageUrl);
+      return this.area.set({
+        [originStoreKey]: data
+      });
+    },
+
+    remove(pageUrl) {
+      const originStoreKey = this.getOriginStoreKey(pageUrl);
+      return this.area.remove([originStoreKey]);
+    },
+  },
+
+  init() {
+    browser.tabs.onActivated.addListener((info) => {
+      browser.tabs.get(info.tabId).then((tab) => {
+        this.calculateContextMenu(tab);
+      }).catch((e) => {
+        throw e;
+      });
+    });
+
+    browser.windows.onFocusChanged.addListener(() => {
+      browser.tabs.query({active: true}).then((tabs) => {
+        if (tabs && tabs[0]) {
+          this.calculateContextMenu(tabs[0]);
+        }
+      }).catch((e) => {
+        throw e;
+      });
+    });
+
+    browser.runtime.onMessage.addListener((neverAskMessage) => {
+      const pageUrl = neverAskMessage.pageUrl;
+      if (neverAskMessage.neverAsk === true) {
+        // If we have existing data and for some reason it hasn't been deleted etc lets update it
+        this.storageArea.get(pageUrl).then((originSettings) => {
+          if (originSettings) {
+            originSettings.neverAsk = true;
+            this.storageArea.set(pageUrl, originSettings);
+          }
+        }).catch((e) => {
+          throw e;
+        });
+      }
+    });
+
+    browser.contextMenus.onClicked.addListener((info, tab) => {
+      const userContextId = this.getUserContextIdFromCookieStore(tab);
+      // Mapping ${URL(info.pageUrl).origin} to ${userContextId}
+      if (userContextId) {
+        let actionName;
+        let storageAction;
+        if (info.menuItemId === this.MENU_ASSIGN_ID) {
+          actionName = "added";
+          storageAction = this.storageArea.set(info.pageUrl, {
+            userContextId,
+            neverAsk: false
+          });
+        } else {
+          actionName = "removed";
+          storageAction = this.storageArea.remove(info.pageUrl);
+        }
+        storageAction.then(() => {
+          const pageUrl = new window.URL(info.pageUrl);
+          browser.notifications.create({
+            type: "basic",
+            title: "Containers",
+            message: `Successfully ${actionName} ${pageUrl.origin} to always open in this container`,
```

Comment:
What if we just remove the `${pageUrl.origin}` completely? A significant origin string overflows the notification interface, and the important feedback here seems to be the action. Can we assume knows the origin they just assigned, since they just assigned it?